# Lab | Pandas

In this lab, we will be working with the customer data from an insurance company, which can be found in the CSV file located at the following link: https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file1.csv

The data includes information such as customer ID, state, gender, education, income, and other variables that can be used to perform various analyses.

Throughout the lab, we will be using the pandas library in Python to manipulate and analyze the data. Pandas is a powerful library that provides various data manipulation and analysis tools, including the ability to load and manipulate data from a variety of sources, including CSV files.

### Data Description

- Customer - Customer ID

- ST - State where customers live

- Gender - Gender of the customer

- Education - Background education of customers 

- Customer Lifetime Value - Customer lifetime value(CLV) is the total revenue the client will derive from their entire relationship with a customer. In other words, is the predicted or calculated value of a customer over their entire duration as a policyholder with the insurance company. It is an estimation of the net profit that the insurance company expects to generate from a customer throughout their relationship with the company. Customer Lifetime Value takes into account factors such as the duration of the customer's policy, premium payments, claim history, renewal likelihood, and potential additional services or products the customer may purchase. It helps insurers assess the long-term profitability and value associated with retaining a particular customer.

- Income - Customers income

- Monthly Premium Auto - Amount of money the customer pays on a monthly basis as a premium for their auto insurance coverage. It represents the recurring cost that the insured person must pay to maintain their insurance policy and receive coverage for potential damages, accidents, or other covered events related to their vehicle.

- Number of Open Complaints - Number of complaints the customer opened

- Policy Type - There are three type of policies in car insurance (Corporate Auto, Personal Auto, and Special Auto)

- Vehicle Class - Type of vehicle classes that customers have Two-Door Car, Four-Door Car SUV, Luxury SUV, Sports Car, and Luxury Car

- Total Claim Amount - the sum of all claims made by the customer. It represents the total monetary value of all approved claims for incidents such as accidents, theft, vandalism, or other covered events.


External Resources: https://towardsdatascience.com/filtering-data-frames-in-pandas-b570b1f834b9

## Challenge 1: Understanding the data

In this challenge, you will use pandas to explore a given dataset. Your task is to gain a deep understanding of the data by analyzing its characteristics, dimensions, and statistical properties.

- Identify the dimensions of the dataset by determining the number of rows and columns it contains.
- Determine the data types of each column and evaluate whether they are appropriate for the nature of the variable. You should also provide suggestions for fixing any incorrect data types.
- Identify the number of unique values for each column and determine which columns appear to be categorical. You should also describe the unique values of each categorical column and the range of values for numerical columns, and give your insights.
- Compute summary statistics such as mean, median, mode, standard deviation, and quartiles to understand the central tendency and distribution of the data for numerical columns. You should also provide your conclusions based on these summary statistics.
- Compute summary statistics for categorical columns and providing your conclusions based on these statistics.

In [38]:
# Your code here

# Load the dataset from an online source
import pandas as pd
url = 'https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file1.csv'
data_insurance = pd.read_csv(url)

# 1- Identify the dimensions of the dataset by determining the number of rows and columns it contains.
data_insurance.shape



(4008, 11)

In [9]:
# 2- Determine the data types of each column and evaluate whether they are appropriate for the nature of the variable. You should also provide suggestions for fixing any incorrect data types.
data_insurance.dtypes

# Suggestions: "Customer Lifetime Value" should be a float, as it is a number and the "Number of Open Complaints" should be a integer


Customer                         str
ST                               str
GENDER                           str
Education                        str
Customer Lifetime Value          str
Income                       float64
Monthly Premium Auto         float64
Number of Open Complaints        str
Policy Type                      str
Vehicle Class                    str
Total Claim Amount           float64
dtype: object

In [20]:
# 3- Identify the number of unique values for each column and determine which columns appear to be categorical. You should also describe the unique values of each categorical column and the rabnge of values for numerical columns, and give your insights.

data_insurance.value_counts()

# I did manual ==> Categorical: Customer, ST, Gender, Education, Policy Type, Vehicule Class

Customer  ST          GENDER  Education             Customer Lifetime Value  Income   Monthly Premium Auto  Number of Open Complaints  Policy Type     Vehicle Class  Total Claim Amount
QZ44356   Arizona     F       Bachelor              697953.59%               0.0      94.0                  1/0/00                     Personal Auto   Four-Door Car  1131.464935           1
AI49188   Nevada      F       Bachelor              1288743.17%              48767.0  108.0                 1/0/00                     Personal Auto   Two-Door Car   566.472247            1
WW63253   California  M       Bachelor              764586.18%               0.0      106.0                 1/0/00                     Corporate Auto  SUV            529.881344            1
GA49547   Washington  M       High School or Below  536307.65%               36357.0  68.0                  1/0/00                     Personal Auto   Four-Door Car  17.269323             1
OC83172   Oregon      F       Bachelor              825

In [19]:
data_insurance.nunique()

Customer                     1071
ST                              8
GENDER                          5
Education                       6
Customer Lifetime Value      1027
Income                        774
Monthly Premium Auto          132
Number of Open Complaints       6
Policy Type                     3
Vehicle Class                   6
Total Claim Amount            761
dtype: int64

In [27]:
categorical_cols = data_insurance.select_dtypes(include="object").columns
numerical_cols = data_insurance.select_dtypes(include=["int64", "float64"]).columns

print(categorical_cols)
print(numerical_cols)



Index(['Customer', 'ST', 'GENDER', 'Education', 'Customer Lifetime Value',
       'Number of Open Complaints', 'Policy Type', 'Vehicle Class'],
      dtype='str')
Index(['Income', 'Monthly Premium Auto', 'Total Claim Amount'], dtype='str')


C:\Users\ASUS\AppData\Local\Temp\ipykernel_1064\1827316068.py:1: Pandas4Warning: For backward compatibility, 'str' dtypes are included by select_dtypes when 'object' dtype is specified. This behavior is deprecated and will be removed in a future version. Explicitly pass 'str' to `include` to select them, or to `exclude` to remove them and silence this warning.
See https://pandas.pydata.org/docs/user_guide/migration-3-strings.html#string-migration-select-dtypes for details on how to write code that works with pandas 2 and 3.
  categorical_cols = data_insurance.select_dtypes(include="object").columns


In [29]:
for col in categorical_cols: # For each categorical column, look at the unique categories:
    print(f"\n{col}")
    print(data_insurance[col].value_counts())


Customer
Customer
RB50392    1
QZ44356    1
AI49188    1
WW63253    1
GA49547    1
          ..
TM65736    1
VJ51327    1
GS98873    1
CW49887    1
MY31220    1
Name: count, Length: 1071, dtype: int64

ST
ST
Oregon        320
California    211
Arizona       186
Cali          120
Nevada         98
Washington     81
WA             30
AZ             25
Name: count, dtype: int64

GENDER
GENDER
F         457
M         413
Male       39
female     28
Femal      17
Name: count, dtype: int64

Education
Education
Bachelor                324
College                 313
High School or Below    296
Master                   94
Doctor                   37
Bachelors                 7
Name: count, dtype: int64

Customer Lifetime Value
Customer Lifetime Value
445811.34%     4
251459.20%     4
2412750.40%    3
684615.03%     3
578018.22%     3
              ..
284226.69%     1
305955.03%     1
2031499.76%    1
323912.47%     1
899704.02%     1
Name: count, Length: 1027, dtype: int64

Number of Open Com

In [30]:
data_insurance[numerical_cols].describe()

,Income,Monthly Premium Auto,Total Claim Amount
count,1071.000000,1071.000000,1071.000000
mean,39295.701214,193.234360,404.986909
std,30469.427060,1601.190369,293.027260
min,0.000000,61.000000,0.382107
25%,14072.000000,68.000000,202.157702
50%,36234.000000,83.000000,354.729129
75%,64631.000000,109.500000,532.800000
max,99960.000000,35354.000000,2893.239678


In [31]:
# Number of unique values
print(data_insurance.nunique())

# Categorical columns
categorical_cols = data_insurance.select_dtypes(include='object').columns

# Numerical columns
numerical_cols = data_insurance.select_dtypes(include=['int64', 'float64']).columns

# Unique values in categorical columns
for col in categorical_cols:
    print(f"\n{col}")
    print(data_insurance[col].unique())

# Summary of numerical columns
print(data_insurance[numerical_cols].describe())

Customer                     1071
ST                              8
GENDER                          5
Education                       6
Customer Lifetime Value      1027
Income                        774
Monthly Premium Auto          132
Number of Open Complaints       6
Policy Type                     3
Vehicle Class                   6
Total Claim Amount            761
dtype: int64

Customer
<StringArray>
['RB50392', 'QZ44356', 'AI49188', 'WW63253', 'GA49547', 'OC83172', 'XZ87318',
 'CF85061', 'DY87989', 'BQ94931',
 ...
 'IK12620', 'GQ66762', 'YT69858', 'XD85577', 'TM65736', 'VJ51327', 'GS98873',
 'CW49887', 'MY31220',       nan]
Length: 1072, dtype: str

ST
<StringArray>
['Washington',    'Arizona',     'Nevada', 'California',     'Oregon',
       'Cali',         'AZ',         'WA',          nan]
Length: 9, dtype: str

GENDER
<StringArray>
[nan, 'F', 'M', 'Femal', 'Male', 'female']
Length: 6, dtype: str

Education
<StringArray>
[              'Master',             'Bachelor', 'High 

C:\Users\ASUS\AppData\Local\Temp\ipykernel_1064\470745685.py:5: Pandas4Warning: For backward compatibility, 'str' dtypes are included by select_dtypes when 'object' dtype is specified. This behavior is deprecated and will be removed in a future version. Explicitly pass 'str' to `include` to select them, or to `exclude` to remove them and silence this warning.
See https://pandas.pydata.org/docs/user_guide/migration-3-strings.html#string-migration-select-dtypes for details on how to write code that works with pandas 2 and 3.
  categorical_cols = data_insurance.select_dtypes(include='object').columns


In [ ]:
# My insights:
# There are 1,071 registered customers.
# The gender field should contain only "Female" or "Male", but there are misspellings.
# Customer, ST, Gender, Education, Policy Type, Vehicule Class are categorical variables

In [34]:
# 4- Compute summary statistics such as mean, median, mode, standard deviation, and quartiles to understand the central tendency and distribution of the data for numerical columns. You should also provide your conclusions based on these summary statistics.

# Summary of numerical columns
print(data_insurance[numerical_cols].describe())


             Income  Monthly Premium Auto  Total Claim Amount
count   1071.000000           1071.000000         1071.000000
mean   39295.701214            193.234360          404.986909
std    30469.427060           1601.190369          293.027260
min        0.000000             61.000000            0.382107
25%    14072.000000             68.000000          202.157702
50%    36234.000000             83.000000          354.729129
75%    64631.000000            109.500000          532.800000
max    99960.000000          35354.000000         2893.239678


In [36]:
# 5- - Compute summary statistics for categorical columns and providing your conclusions based on these statistics.

data_insurance.describe(include = "object")

C:\Users\ASUS\AppData\Local\Temp\ipykernel_1064\386924615.py:3: Pandas4Warning: For backward compatibility, 'str' dtypes are included by select_dtypes when 'object' dtype is specified. This behavior is deprecated and will be removed in a future version. Explicitly pass 'str' to `include` to select them, or to `exclude` to remove them and silence this warning.
See https://pandas.pydata.org/docs/user_guide/migration-3-strings.html#string-migration-select-dtypes for details on how to write code that works with pandas 2 and 3.
  data_insurance.describe(include = "object")


,Customer,ST,GENDER,Education,Customer Lifetime Value,Number of Open Complaints,Policy Type,Vehicle Class
count,1071,1071,954,1071,1068,1071,1071,1071
unique,1071,8,5,6,1027,6,3,6
top,RB50392,Oregon,F,Bachelor,445811.34%,1/0/00,Personal Auto,Four-Door Car
freq,1,320,457,324,4,830,780,576


My insights:

1. The Customer column acts as a unique identifier and does not provide analytical value for understanding customer behavior.
2. Oregon is the most represented state in the dataset, suggesting that the customer population is not evenly distributed geographically.
3. The Gender column contains missing values and more categories than expected, indicating potential data quality issues that should be investigated before analysis.
4. Bachelor's degree holders represent the largest educational group in the dataset, suggesting a relatively well-educated customer base.
5. Customer Lifetime Value appears highly variable and may require cleaning and conversion to a numeric format before further analysis.
6. The Number of Open Complaints column appears to contain formatting inconsistencies and may have been imported incorrectly, requiring data cleaning.
7. Personal Auto is by far the most common policy type, accounting for approximately 73% of all policies.
8. Four-Door Cars are the dominant vehicle class in the dataset, representing more than half of all insured vehicles.




## Challenge 2: analyzing the data

### Exercise 1

The marketing team wants to know the top 5 less common customer locations. Create a pandas Series object that contains the customer locations and their frequencies, and then retrieve the top 5 less common locations in ascending order.

In [48]:
# Your code here

location_freq = data_insurance["ST"].value_counts()

top_5_less_common = location_freq.sort_values().head(5)

print(top_5_less_common)

ST
AZ             25
WA             30
Washington     81
Nevada         98
Cali          120
Name: count, dtype: int64


### Exercise 2

The sales team wants to know the total number of policies sold for each type of policy. Create a pandas Series object that contains the policy types and their total number of policies sold, and then retrieve the policy type with the highest number of policies sold.

*Hint:*
- *Using value_counts() method simplifies this analysis.*
- *Futhermore, there is a method that returns the index of the maximum value in a column or row.*


In [49]:
# Your code here

policy_counts = data_insurance["Policy Type"].value_counts()

policy_counts

Policy Type
Personal Auto     780
Corporate Auto    234
Special Auto       57
Name: count, dtype: int64

In [50]:
policy_counts.idxmax()

'Personal Auto'

In [51]:
most_sold_policy = policy_counts.idxmax()
number_sold = policy_counts.max()

print(most_sold_policy)
print(number_sold)

Personal Auto
780


Personal Auto is the most frequently sold policy type, with 780 policies. It represents the majority of policies in the dataset, indicating that individual customers make up the largest segment of the company's customer base.

### Exercise 3

The sales team wants to know if customers with Personal Auto have a lower income than those with Corporate Auto. How does the average income compare between the two policy types?

- Use *loc* to create two dataframes: one containing only Personal Auto policies and one containing only Corporate Auto policies.
- Calculate the average income for each policy.
- Print the results.

In [54]:
# Your code here
personal_auto = data_insurance.loc[
    data_insurance["Policy Type"] == "Personal Auto"
]

corporate_auto = data_insurance.loc[
    data_insurance["Policy Type"] == "Corporate Auto"
]

In [55]:
personal_income_avg = personal_auto["Income"].mean()
corporate_income_avg = corporate_auto["Income"].mean()

In [56]:
print("Average income - Personal Auto:", personal_income_avg)
print("Average income - Corporate Auto:", corporate_income_avg)

Average income - Personal Auto: 38180.69871794872
Average income - Corporate Auto: 41390.31196581197


### Bonus: Exercise 4


Your goal is to identify customers with a high policy claim amount.

Instructions:

- Review again the statistics for total claim amount to gain an understanding of the data.
- To identify potential areas for improving customer retention and profitability, we want to focus on customers with a high policy claim amount. Consider customers with a high policy claim amount to be those in the top 25% of the total claim amount. Create a pandas DataFrame object that contains information about customers with a policy claim amount greater than the 75th percentile.
- Use DataFrame methods to calculate summary statistics about the high policy claim amount data. 

*Note: When analyzing data, we often want to focus on certain groups of values to gain insights. Percentiles are a useful tool to help us define these groups. A percentile is a measure that tells us what percentage of values in a dataset are below a certain value. For example, the 75th percentile represents the value below which 75% of the data falls. Similarly, the 25th percentile represents the value below which 25% of the data falls. When we talk about the top 25%, we are referring to the values that fall above the 75th percentile, which represent the top quarter of the data. On the other hand, when we talk about the bottom 25%, we are referring to the values that fall below the 25th percentile, which represent the bottom quarter of the data. By focusing on these groups, we can identify patterns and trends that may be useful for making decisions and taking action.*

*Hint: look for a method that gives you the percentile or quantile 0.75 and 0.25 for a Pandas Series.*

*Hint 2: check `Boolean selection according to the values of a single column` in https://towardsdatascience.com/filtering-data-frames-in-pandas-b570b1f834b9*

In [ ]:
# Your code here